### PR TITLE
Add AOT support for the EntryPoint property for UnmanagedCallersOnly.

### LINKED ISSUE #44803: [aot] Mono should support UnmanagedCallersOnly EntryPoint attribute
Expected behavior:
  The (wrapper of the) method will be exported with the specified name.
Actual behavior:
  The `EntryPoint` is not used

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/dotnet/runtime/issues/44803.

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -5060,10 +5060,42 @@ MONO_RESTORE_WARNING
...
+				for (j = 0; j < num_named_args; ++j) {
+					if (named_arg_info [j].field && !strcmp (named_arg_info [j].field->name, "EntryPoint")) {
+						named = named_args [j];
+						slen = mono_metadata_decode_value (named, &named);
+						export_name = (char *)g_malloc (slen + 1);
+						memcpy (export_name, named, slen);
+						export_name [slen] = 0;
+					}
+				}
+				g_free (named_args);
```
Comment (lambdageek) [inline src/mono/mono/mini/aot-compiler.c:5090]:
```suggestion
				g_assert (!typed_args);
				g_free (named_args);
```

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/mini/driver.c [hunk 2]
```diff
@@ -452,7 +452,9 @@ method_should_be_regression_tested (MonoMethod *method, gboolean interp)
 		if (!is_ok (error))
 			continue;
 
-		char *utf8_str = (char*)(void*)typed_args[0]; //this points into image memory that is constant
+		gpointer arg = typed_args [0];
+		mono_metadata_decode_value (arg, (const char**)&arg);
+		char *utf8_str = (char*)(void*)arg; //this points into image memory that is constant
```
Comment (lambdageek) [inline src/mono/mono/mini/driver.c:457]:
```suggestion
		const char *arg = (const char*)typed_args [0];
		mono_metadata_decode_value (arg, &arg);
		char *utf8_str = (char*)arg; //this points into image memory that is constant
```

(fixes the C++ build in mono/mono)

File: src/mono/mono/mini/aot-compiler.c [hunk 3]
```diff
@@ -5060,10 +5060,42 @@ MONO_RESTORE_WARNING
...
 			if (j < cattr->num_attrs) {
-				MonoMethod *wrapper = mono_marshal_get_managed_wrapper (method, NULL, 0, error);
+				MonoCustomAttrEntry *e = &cattr->attrs [j];
+				const char *named;
+				int slen;
+				char *export_name = NULL;
+				MonoMethod *wrapper;
+
+				if (!(method->flags & METHOD_ATTRIBUTE_STATIC)) {
+					g_warning ("AOT restriction: Method '%s' must be static since it is decorated with [UnmanagedCallers].",
```
Comment (marek-safar) [inline src/mono/mono/mini/aot-compiler.c:5070]: Is this permanent limitation?

Comment (lambdageek) [reply src/mono/mono/mini/aot-compiler.c:5070]: It's part of the [spec of `UnmanagedCallersOnlyAttribute`](https://github.com/dotnet/csharplang/blob/7cf8d52977233ee1995e48c7028723fa433460ce/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute)
> It is an error to apply the attribute to anything other than an ordinary static method or ordinary static local function. The C# compiler will mark any non-static or static non-ordinary methods imported from metadata with this attribute as unsupported by the language.

Comment (marek-safar) [reply src/mono/mono/mini/aot-compiler.c:5070]: Why do we check only for this restriction then and shouldn't this be TLE or something like that?

Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:5070]: Its just a copy paste from the MonoPInvokeCallback case.

**Pull request merged**